<launch>

  <arg name="ouster_ns" default="ouster" doc="Override the default namespace of all ouster nodes"/>
  <arg name="sensor_hostname" doc="The hostname or IP in dotted decimal form of the sensor"/>
  <arg name="udp_dest" default=" " doc="The hostname or IP where the sensor will send data packets"/>
  <arg name="retry_configuration" default="false" doc="True to re-attempt unsuccessful sensor configuration routines e.g. if the sensor is disconnected"/>
  <arg name="lidar_port" default="0" doc="The port to which the sensor should send lidar data"/>
  <arg name="imu_port" default="0" doc="The port to which the sensor should send imu data"/>
  <arg name="udp_profile_lidar" default=" " doc="The lidar packet profile; possible values: {
    LEGACY,
    RNG19_RFL8_SIG16_NIR16_DUAL,
    RNG19_RFL8_SIG16_NIR16,
    RNG15_RFL8_NIR8
    }"/>
  <arg name="lidar_mode" default=" " doc="The resolution and rate; possible values: {
    512x10,
    512x20,
    1024x10,
    1024x20,
    2048x10,
    4096x5
    }"/>
  <arg name="timestamp_mode" default=" " doc="The method used to timestamp measurements; possible values: {
    TIME_FROM_INTERNAL_OSC,
    TIME_FROM_SYNC_PULSE_IN,
    TIME_FROM_PTP_1588,
    TIME_FROM_ROS_TIME
    }"/>
  <arg name="metadata" default=" " doc="The path to write metadata file when receiving sensor data"/>
  <arg name="viz" default="true" doc="Whether to run an rviz instance"/>
  <arg name="rviz_config" default="$(find ouster_ros)/config/viz.rviz" doc="The optional rviz config file"/>
  <arg name="tf_prefix" default=" " doc="The namespace for tf transforms"/>

  <arg name="no_bond" default="false"
    doc="The request no bond setup when nodelets are created"/>
  <arg if="$(arg no_bond)" name="_no_bond" value="--no-bond"/>
  <arg unless="$(arg no_bond)" name="_no_bond" value="" />

  <group ns="$(arg ouster_ns)">
    <node pkg="nodelet" type="nodelet" name="os_nodelet_mgr"
      output="screen" required="true"
      launch-prefix="bash -c 'sleep 2; $0 $@' "
      args="manager"/>
  </group>

  <group ns="$(arg ouster_ns)">
    <node pkg="nodelet" type="nodelet" name="os_node"
      output="screen" required="true"
      args="load nodelets_os/OusterSensor os_nodelet_mgr $(arg _no_bond)">
      <param name="~/sensor_hostname" type="str" value="$(arg sensor_hostname)"/>
      <param name="~/udp_dest" type="str" value="$(arg udp_dest)"/>
      <param name="~/retry_configuration" type="bool" value="$(arg retry_configuration)"/>
      <param name="~/lidar_port" type="int" value="$(arg lidar_port)"/>
      <param name="~/imu_port" type="int" value="$(arg imu_port)"/>
      <param name="~/udp_profile_lidar" type="str" value="$(arg udp_profile_lidar)"/>
      <param name="~/lidar_mode" type="str" value="$(arg lidar_mode)"/>
      <param name="~/timestamp_mode" type="str" value="$(arg timestamp_mode)"/>
      <param name="~/metadata" type="str" value="$(arg metadata)"/>
    </node>
  </group>

  <include file="$(find ouster_ros)/launch/common.launch">
    <arg name="ouster_ns" value="$(arg ouster_ns)"/>
    <arg name="viz" value="$(arg viz)"/>
    <arg name="rviz_config" value="$(arg rviz_config)"/>
    <arg name="tf_prefix" value="$(arg tf_prefix)"/>
    <arg name="timestamp_mode" value="$(arg timestamp_mode)"/>
    <arg name="_no_bond" value="$(arg _no_bond)"/>
  </include>

</launch>
